### PR TITLE
 Adding support for secondary interfaces

### PR DESCRIPTION
**Pull request by burhan20**

### CODE REVIEW

File: app/Dockerfile [hunk 1]
```diff
@@ -2,6 +2,7 @@ FROM python:3.10-slim
 WORKDIR /app
 ENV PATH "$PATH:/app"
 COPY requirements.txt requirements.txt
+RUN echo "deb http://deb.debian.org/debian bookworm main" > /etc/apt/sources.list
```
Comment (abhis2112) [inline app/Dockerfile]: why this change is needed ?

File: helm/charts/neo4j-vkaci/templates/neo4j-service-account.yaml [hunk 2]
```diff
@@ -36,3 +36,4 @@ roleRef:
   name: {{ .Release.Name }}-service-reader # this must match the name of the Role or ClusterRole you wish to bind to
   apiGroup: rbac.authorization.k8s.io
 {{- end }}
+
```
Comment (abhis2112) [inline helm/charts/neo4j-vkaci/templates/neo4j-service-account.yaml]: remove extra line

File: app/Dockerfile (hunk 1 above)
Comment (burhan20) [reply app/Dockerfile]: The build is failing actually. It is not able to find /etc/apt/sources.list. This is not final yet, checking why build is failing

File: helm/templates/clusterrole.yaml [hunk 3]
```diff
@@ -29,4 +29,32 @@ subjects:
 roleRef:
   kind: ClusterRole
   name: {{ include "vkaci.fullname" . }}-pods-list
+  apiGroup: rbac.authorization.k8s.io
+---
```
Comment (abhis2112) [inline helm/templates/clusterrole.yaml]: ideally should be a separate file naming role.yaml which includes both role and rolebinding  as here in clusterrole.yaml

File: unit_test.py [hunk 4]
```diff
@@ -223,13 +224,13 @@ def test_valid_topology(self):
...
+                                                   'kube-router-xfgr': {'ip': '192.168.1.2', 'primary_iface': '', 'ns': 'kube-system', 'labels': {}, 'other_ifaces': {}, 'annotations': {}}},
                                           'bgp_peers': {'leaf-204': {'prefix_count': 2}}, 'neighbours': {'esxi4.cam.ciscolabs.com':
                                                                                                          {'switches': {'leaf-204': {'vmxnic1-eth1/1'}}, 'Description': 'VMware version 123'}},
-                                          'labels': {'app': 'redis'}, 'mac': 'MOCKMO1C'}},
-                    'services': {'appx': [{'name': 'example service', 'cluster_ip': '192.168.25.5', 'external_i_ps': ['192.168.5.1'],'load_balancer_ip': '192.168.5.2','ns':'appx',
-                                           'labels': {'app': 'guestbook'}}]}}
+                                          'labels': {'app': 'redis'}, 'node_leaf_sec_iface_conn': [], 'node_pod_sec_iface_conn': [], 'node_leaf_ter_iface_conn': [], 'node_pod_ter_iface_conn': [], 'node_leaf_all_iface_conn': [], 'mac': 'MOCKMO1C'}},
```
Comment (abhis2112) [inline unit_test.py:231]: we have added null values everywhere . Is this correct? Should be some value right

Comment (abhis2112) [reply unit_test.py:231]: same comment wherever we are adding null values

Comment (burhan20) [reply unit_test.py:231]: We have added these additional parameters in our structure. We had to pass these parameters to the already present testcases as well otherwise the testcases would fail.

File: app/graph.py [hunk 5]
```diff
@@ -36,7 +37,7 @@ def __init__(self, dict_env:dict = None):
         else:
             self.apic_ip = []
 
-        self.aciMetaFilePath = self.enviro().get("ACI_META_FILE", "/app/aci-meta/aci-meta.json")
+        self.aciMetaFilePath = self.enviro().get("ACI_META_FILE", "/root/.aci-meta/aci-meta.json")
```
Comment (camrossi) [inline app/graph.py]: I had to change from `/root` to `/app` to make it work across OpenShift and Upstream Kubernetes and I fear this might break things again. Need to test it in my own environment

File: app/Dockerfile (hunk 1 above)
Comment (camrossi) [reply app/Dockerfile]: Not sure why is happening but started to fail also for my main branch. I just removed the line altogether and works now

File: app/graph.py (hunk 5 above)
Comment (camrossi) [reply app/graph.py]: Indeed this breaks compatibility with Upstream K8s
```
pi@pi-k8s-1:~$ kubectl  logs vkaci-7548c967f8-slds6 -c init-vkaci
2023-06-15 17:05:40,284 __main__ INFO Loading ACI Metadata
Traceback (most recent call last):
  File "/app/init.py", line 29, in <module>
    file = open('/root/.aci-meta/aci-meta.json','wb')
FileNotFoundError: [Errno 2] No such file or directory: '/root/.aci-meta/aci-meta.json'
```

Just revert this back to `/app/aci-meta/aci-meta.json`

File: app/graph.py [hunk 6]
```diff
@@ -470,23 +487,108 @@ def update(self):
...
+                                    if val["ips"][0] != pod_ip:
+                                        iface_name = str(val["name"].split('/')[-1])
+                                        pod_iface = str(val["interface"])
+                                        pods[pod_name]['other_ifaces'][iface_name] = pod_iface
+                                    else:
+                                        pods[pod_name]['primary_iface'] = str(val["interface"])
+            except Exception as e:
+                logger.error(f"Error processing pod: {i.metadata.name}. Error: {str(e)}")
+
+        cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
```
Comment (camrossi) [inline app/graph.py]: This resource might not exists, at the moment we crash if is not present:
```
kubernetes.client.exceptions.ApiException: (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Audit-Id': '3d0a3afb-ef4a-4163-9341-025e45d0ca3c', 'Cache-Control': 'no-cache, private', 'Content-Type': 'text/plain; charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': 'c3d0b261-bb0f-4204-ba5a-c013697f1875', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'e2fd79d1-4c6d-4bf6-86a8-eea82de11102', 'Date': 'Thu, 15 Jun 2023 17:43:07 GMT', 'Content-Length': '19'})
HTTP response body: 404 page not found
```
we should handle this gracefully as not all cluster where VKACI runs have SR-IOV

File: app/template/assets/cui-2.0.0/js/vkaci.js [hunk 7]
```diff
@@ -16,118 +13,160 @@ function neo_viz_config(showPodName, container, cypher, seed = null) {
...
             },
-
             stabilization: {
-                iterations: 400, // CHANGEME: If want different stabilisation,
-                fit: true
-            }
+                iterations: 400,
+                fit: true,
+            },
+            dragView: true
```
Comment (camrossi) [inline app/template/assets/cui-2.0.0/js/vkaci.js]: In my environment this leads to a crash:
```
Unknown option detected: "dragView" in 

options = {
  physics: {
    dragView
  }
}
```

Not sure why, do you use a different neo4j version ?

File: app/graph.py [hunk 8]
```diff
@@ -626,30 +728,78 @@ def __init__(self, env: VkaciEnvVariables, topology: VkaciBuilTopology) -> None:
 
     # Build query.
     query = """
```
Comment (camrossi) [inline app/graph.py:744]: I think all the query need to be re-written to account the fact that not all cluster will have the `fabricattachment` CR 
Currently if I run this code on a cluster without the CR no visualisation happens

File: app/template/index.html [hunk 9]
```diff
@@ -136,6 +136,9 @@
                               <a id="WithoutBgpPeers" onclick="draw_without_bgp_peers()">Without Bgp Peers</a>
                               <a id="WithoutPods" onclick="draw_without_pods()">Without Pods</a>
                               <a id="OnlyBgpPeers" onclick="draw_only_bgp_peers()">BGP Peers</a>
+                              <a id="OnlyPrimarylinks" onclick="draw_only_primary_links()">Primary Links</a>
```
Comment (camrossi) [inline app/template/index.html]: These new 3 views should only be displayed if the CR `fabricattachment` is present. 
We need to support also clusters without this CR as we have customers using this tool unrelated to SR-IOV

File: app/Dockerfile (hunk 1 above)
Comment (burhan20) [reply app/Dockerfile]: reverted these changes

File: app/template/index.html [hunk 10]
```diff
@@ -133,9 +133,18 @@
                            <button class="btn btn--white btn--dropdown selected view">Different Views</button>
                            <div id="selected_views" class="dropdown__menu dropdown__menu--openleft">
                               <a id="PodsAndNodes" onclick="draw_pods_and_nodes()">Pods & Nodes</a>
-                              <a id="WithoutBgpPeers" onclick="draw_without_bgp_peers()">Without Bgp Peers</a>
+                              {% if not openshift %}
```
Comment (camrossi) [inline app/template/index.html]: I fear the logic is not correct or at minimum to wording is wrong.
We already support OpenShift (without SR-IOV or MacVlan)

What you are adding is support for SR-IOV and MacVALN so I think what you should check is if the new CR is present then you show the menus.
Having a `if not openshift` kinda check is confusing

File: app/graph.py [hunk 11]
```diff
@@ -470,23 +488,111 @@ def update(self):
...
+                                        pod_iface = str(val["interface"])
+                                        pods[pod_name]['other_ifaces'][iface_name] = pod_iface
+                                    else:
+                                        pods[pod_name]['primary_iface'] = str(val["interface"])
+            except Exception as e:
+                logger.error(f"Error processing pod: {i.metadata.name}. Error: {str(e)}")
+
+        cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
+
+        if cr:
```
Comment (abhis2112) [inline app/graph.py]: you need to check if len(cr) == 0 as it is a list

Comment (abhis2112) [reply app/graph.py]: I am not able to view any topology even when it is false. Can you check the logic again here

File: app/graph.py [hunk 12]
```diff
@@ -470,23 +493,112 @@ def update(self):
...
+                        if aciTopology is not None and i["spec"]["nodeName"] == nodeName:
+                            for iface_name, link in i["spec"]["aciTopology"].items():
+                                if "sriov" in i["metadata"]["name"]:
+                                    iface_name = "PF-" + iface_name
+                                fabricLinks = link.get("fabricLink", [])
+                                for fabricLink in fabricLinks:
+                                    fabricLinkSplit = fabricLink.split("/")
+                                    switch_name = fabricLinkSplit[2].replace("node", "leaf")
+                                    switch_interface = fabricLink[fabricLink.rfind('[') + 1: fabricLink.rfind(']')]
+                                    if "sriov" in i["metadata"]["name"]:
```
Comment (abhis2112) [inline app/graph.py]: we need to come up with a more robust solution that how we want to check for the type. The name might not be enough. Will discuss in the group want we want to select a type - a name, or an annotation

File: app/graph.py (hunk 6 above)
Comment (abhis2112) [reply app/graph.py]: @burhan20 i think with latest changes we expect CRD to be present which might not be the case . We need to add try/except  block. Something like below:
```
     try:
            self.custom_obj.get_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
            resource_exists = True
        except client.exceptions.ApiException as e:
            if e.status == 404:
                resource_exists = False
            else:
                raise

        # If the custom resource exists, list objects
        if resource_exists:
            cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1",
                                                               namespace="aci-containers-system",
                                                               plural="nodefabricnetworkattachments")
            if cr.get("items"):
                for nodeName in self.topology['nodes']:
                    try:
                        for i in cr.get("items"):
                            aciTopology = i["spec"].get("aciTopology")
                            if aciTopology is not None and i["spec"]["nodeName"] == nodeName:
                                for iface_name, link in i["spec"]["aciTopology"].items():
                                    if "sriov" in i["metadata"]["name"]:
                                        iface_name = "PF-" + iface_name
                                    fabricLinks = link.get("fabricLink", [])
                                    for fabricLink in fabricLinks:
                                        fabricLinkSplit = fabricLink.split("/")
                                        switch_name = fabricLinkSplit[2].replace("node", "leaf")
                                        switch_interface = fabricLink[fabricLink.rfind('[') + 1: fabricLink.rfind(']')]
                                        if "sriov" in i["metadata"]["name"]:
                                            self.sriov = True
                                            self.topology['nodes'][nodeName]['node_leaf_sec_iface_conn'].append({
                                                'switch_name': switch_name,
                                                'switch_interface': switch_interface,
                                                'node_iface': iface_name
                                            })
                                        else:
                                            self.macvlan = True
                                            self.topology['nodes'][nodeName]['node_leaf_ter_iface_conn'].append({
                                                'switch_name': switch_name,
                                                'switch_interface': switch_interface,
                                                'node_iface': iface_name
                                            })

                                    pods = link.get("pods", [])
                                    for pod in pods:
                                        node_iface = pod.get("localIface")
                                        if "sriov" in i["metadata"]["name"]:
                                            node_iface = "VF-" + node_iface
                                        pod_name = pod.get("podRef")["name"]
                                        network_ref = i["spec"].get("networkRef")
                                        node_network = network_ref["name"]
                                        if "sriov" in i["metadata"]["name"]:
                                            self.topology['nodes'][nodeName]['node_pod_sec_iface_conn'].append({
                                                'node_iface': node_iface,
                                                'pod_name': pod_name,
                                                'node_network': node_network,
                                                'pod_iface': self.topology['nodes'][nodeName]['pods'][pod_name][
                                                    'other_ifaces'].get(node_network, "")
                                            })
                                        else:
                                            self.topology['nodes'][nodeName]['node_pod_ter_iface_conn'].append({
                                                'node_iface': node_iface,
                                                'pod_name': pod_name,
                                                'node_network': node_network,
                                                'pod_iface': self.topology['nodes'][nodeName]['pods'][pod_name][
                                                    'other_ifaces'].get(node_network, "")
                                            })

                        self.topology['nodes'][nodeName]['node_leaf_all_iface_conn'].extend(
                            self.topology['nodes'][nodeName]['node_leaf_sec_iface_conn'])
                        self.topology['nodes'][nodeName]['node_leaf_all_iface_conn'].extend(
                            self.topology['nodes'][nodeName]['node_leaf_ter_iface_conn'])

                    except Exception as e:
                        logger.error(f"Error processing node: {nodeName}. Error: {str(e)}")

Comment (camrossi) [reply app/graph.py]: This works fine in my lab:

```
        try:
            cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
                    ###all the code###
                    except Exception as e:
                        logger.error(f"Error processing node: {nodeName}. Error: {str(e)}")
        except Exception as e:
            if e.status == 404:
                logger.info(f"CRD nodefabricnetworkattachments not detected, SR-IOV/MACVLAN topology support is disabled")
            else:
                logger.error(f"Unexpected error processing list_namespaced_custom_object. Error: {str(e)}")

File: app/template/index.html (hunk 10 above)
Comment (burhan20) [reply app/template/index.html]: Cam, as per the latest changes, we are showing Primary, SRIOV, Macvlan links in dropdown when CR is present.
However, BGP Peers and Without BGP Peers menu we are showing if not openshift case.
Is it the right way to do? Could you please suggest what needs to be done.

File: helm/charts/neo4j-vkaci/values.yaml [hunk 13]
```diff
@@ -214,7 +214,7 @@ config:
 # securityContext defines privilege and access control settings for a Pod or Container. Making sure that we dont run Neo4j as root user.
 securityContext:
   runAsNonRoot: true
-  runAsUser: 1000700000
```
Comment (abhis2112) [inline helm/charts/neo4j-vkaci/values.yaml]: is this change required

File: helm/values.yaml [hunk 14]
```diff
@@ -45,7 +45,7 @@ podSecurityContext: {}
 #This ire required for OpenShift to run the container as non-root 
 securityContext:
   runAsNonRoot: true
-  runAsUser: 1000700000
```
Comment (abhis2112) [inline helm/values.yaml]: same here

File: helm/charts/neo4j-vkaci/values.yaml (hunk 13 above)
Comment (burhan20) [reply helm/charts/neo4j-vkaci/values.yaml]: I will revert this. This value is different for each setup
